### PR TITLE
Rebuild cache observability on Koriym.SemanticLogger

### PR DESCRIPTION
**Pull request by koriym**
## Summary

Rebuilds cache observability on [Koriym.SemanticLogger](https://github.com/koriym/Koriym.SemanticLogger): the cache create/invalidate log becomes an open/event/close tree whose nesting **is** the embed/dependency structure (a parent's embedded children nest under it). The legacy flat `RepositoryLogger` is deprecated but stays bound for BC and receives no internal events.

## Included

- SemanticLogger-based logging at the five AOP boundaries: `CacheInterceptor` / `AbstractDonutCacheInterceptor` (GET) and `CommandInterceptor` / `RefreshInterceptor` / `DonutCommandInterceptor` (command).
- `SafeSemanticLogger` best-effort decorator (a logging failure can never break a cache read/write), `NullSemanticLogger` zero-cost no-op default, bound via `SafeSemanticLoggerProvider`.
- Typed `AbstractContext` subclasses in `src/Log/Context/` with per-context JSON Schemas in `docs/schemas/context/`.
- `invalidate` context records self-describing status words: `roPool`/`etagPool` (`invalidated`|`failed`), `cdn` (`purged`|`failed`), plus `durationMs`. The CDN purge is best-effort and no longer fails local invalidation on outage.
- Manual (direct, non-AOP) top-level `put()` and `invalidateTags()` are rooted in `manual_store` / `manual_invalidate` scopes so their events are not dropped at flush, mirroring the existing `manual_purge` treatment.
- Adds runtime dependency `koriym/semantic-logger ^0.8.0`; `vendor/bin/stree` renders the cache log as a readable tree (`demo/run-dependency.php`, `demo/run-donut.php`).

## Not included (deliberate)

- The multi-embed cache dependency fix and diamond tests already shipped in 1.16.1 (#177).
- A refresh of `docs/llms.txt` / `docs/llms-full.txt` (still reference the old flat log) is left for a separate docs PR.

## Verification

- phpcs (PSR-12) clean; PHPStan (max) and Psalm (level 1) report no errors.
- phpunit: 141 tests, 270 assertions, OK.
- `demo/run-dependency.php` and `demo/run-donut.php` render the SemanticLogger tree (exit 0).

## Commits

1. Rebuild cache observability on Koriym.SemanticLogger
2. Log manual (top-level) put() and invalidateTags() in dedicated scopes

### CODE REVIEW

**Comment by koriym** [comment]
@coderabbitai review

**Comment by koriym** [comment]
@coderabbitai Addressed in df8c8b6:

- **PHPMD `UnusedFormalParameter`**: added `@SuppressWarnings("PHPMD.UnusedFormalParameter")` on `NullSemanticLogger` (class-level) and `SafeSemanticLogger::__unserialize`. Quoted so PHPStan's phpDoc parser accepts the dotted rule name. Preferred over scattering `(void)` casts through each no-op body.
- **Coverage**: the no-op `NullSemanticLogger` methods and the best-effort catch blocks in `SafeSemanticLogger::event()/close()` were uncovered. Added `NullSemanticLoggerTest` and two `SafeSemanticLogger` failure tests, restoring 100% on the changed code.
- **Nitpick (StructuredRepositoryLoggerInterface)**: merged the two stacked docblocks into one.
- **Nitpick (SemanticLogTreeTrait)**: `unlink()` the temp file in the `finally`.

Not changed, with reasons:

- **`instanceof SafeSemanticLogger` + `isTopLevel()`**: kept for consistency with the existing `purge()` path, which already uses the same check. Introducing a `TopLevelManualScopeLogger` interface for only the new call sites would be inconsistent; a holistic refactor across all manual-scope sites can be a separate change.
- **`psalm.xml` directory-wide `InvalidClassConstantType` suppression**: every `Log/Context` class shares the same untyped-constant override from `AbstractContext`, so the directory-level suppression is intentional; per-class `@psalm-suppress` across ~18 classes would add more noise than it removes.

**Comment by koriym** [comment]
@coderabbitai Follow-up (force-pushed c5f2099): for the codecov gap, `NullSemanticLogger` is a pure no-op null object, so its methods are now marked `@codeCoverageIgnore` (matching the existing `/** @codeCoverageIgnore */` usage in the codebase) instead of a no-op test. The `SafeSemanticLogger::event()/close()` resilience paths keep their real failure tests.

**Comment by koriym** [comment]
@coderabbitai Recording the reasoning for the design points raised in review (an out-of-band adversarial pass surfaced the same three):

1. **CDN purge — now fail-closed** (eb337fb). `ResourceStorage::invalidateTags()` previously swallowed a purger exception and still reported success, which could silently leave stale CDN content after a write. Restored the 1.x fail-closed behavior: the local pools are invalidated, the outcome is logged as `cdn: failed`, then the purge exception propagates. Logging stays a side-channel and never masks the failure.

2. **The logger constructor type change is not a BC break.** The constructor is not part of the public contract — neither `QueryRepositoryInterface` nor `ResourceStorageInterface` declares it, and consumers use the interfaces via DI (the module binds `SemanticLoggerInterface`). So no compatibility bridge is added; the `RepositoryLoggerInterface` binding is retained as an intentional, inert deprecation.

3. **Singleton logger under concurrent runtimes — documented + deferred to #179.** The singleton binding is pre-existing (1.x bound the logger as a singleton too). Under Swoole/RoadRunner coroutines an interleaved `open`/`close` can drop a request's log. However: cache behavior is unaffected (logging is best-effort), the default PHP-FPM deployment is unaffected, and the robust fix (request/coroutine-scoped logger) belongs with the host flush-lifecycle. Documented in `tests/CACHE_DEPENDENCY_TESTS.md` and tracked in #179; intentionally out of scope here to keep this PR focused.

**Comment by koriym** [comment]
@coderabbitai The status is still "request," but is it okay to mark them all as "approved"?

**Comment by koriym** [comment]
@coderabbitai review

Since your last pass this branch gained: the donut Last-Modified/Age fix, negative-lifetime clamp, in-band recording of donut write failures (cache_error{write}), cdn_headers (literal CDN-facing response headers incl. flavor defaults and Akamai's Edge-Cache-Tag rename), conditional_request (the 304 decision at the transfer boundary), manual_invalidate_result (invalidate detail stays an event for tag correlation), a #[Purge] demo + annotation closure guard, order-independent tests, and docs/what-the-log-proves.md stating the guarantee and its boundaries. The PR body is now a feature × explainability matrix (EN/JA).

**Comment by koriym** [comment]
@coderabbitai Addressed in 0e229d2.

**Adopted (5 + 1 completed):** the ResourceStorage saveEtag hard-coded UriTag was already fixed in 6a1f5c3, and your comment led to finding the same pattern in DonutRepository's two pre-write cleanup sites — the injected strategy now reaches them too. Also: the test recorder resets depth on flush() like SafeSemanticLogger; llms-full.txt no longer implies per-request flushing isolates concurrent coroutines (sequential loops only, NullSemanticLogger or per-request injector until #179); CACHE_DEPENDENCY_TESTS.md names the unclosed-scope consequence; validate.php PHPDoc widened; PSR-12 brace spaces in ResourceStorageTest.

**Declined:** psalm InvalidClassConstantType scope — the src/Log/Context directory is the precise scope: all 26 context classes share the identical TYPE/SCHEMA_URL const-override pattern, so per-file suppressions add 26 lines of noise for no added safety. SafeSemanticLogger __unserialize $data — the project's PHPMD config passes and the magic method's signature is the language's.

**Stale (reference code this branch no longer has):** src/Log/NullSemanticLogger.php was removed in the 0.9 adaptation (the core NullSemanticLogger is used); CHANGELOG's session count already reads thirteen.

**Comment by koriym** [comment]
@coderabbitai Fixed in dc85adc — you were right, and this one was my triage miss (it sat in the review body, not the inline comments).

Both isNotModified() implementations now close the conditional_request scope on a throwing lookup, recording the established read-outage idiom inside it: cache_error{operation: read} with the request URI when the SAPI array carries one, closed as cache_miss{layer: etag} (cache_error + miss = degraded, lone miss = cold). The exception keeps its pre-existing path — whether the 304 check should degrade to a full response instead is the same behavior decision as the donut write and is tracked for a post-merge issue. Pinned with a throwing ETag pool for both classes; removing either closure fails the suite.

File: psalm.xml [hunk 1]
```diff
@@ -20,5 +20,12 @@
                 <file name="src/RedisDsnProvider.php" />
             </errorLevel>
         </UndefinedClass>
+        <!-- Koriym\SemanticLogger\AbstractContext declares TYPE/SCHEMA_URL = '' (untyped), which Psalm
+             infers as the literal '' type; concrete Context subclasses legitimately set string values. -->
+        <InvalidClassConstantType>
+            <errorLevel type="suppress">
+                <directory name="src/Log/Context" />
+            </errorLevel>
+        </InvalidClassConstantType>
```
[the remark this answers is not in the document]
Comment (koriym) [reply psalm.xml:29]: @coderabbit Not applicable — the directory is the precise scope: all 26 context classes share the identical TYPE/SCHEMA_URL override pattern, so per-file suppressions add 26 lines for no added safety.

File: src/Log/NullSemanticLogger.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    #[Override]
+    public function event(AbstractContext $context): void
+    {
+    }
+
+    #[Override]
+    public function close(AbstractContext $context, string $openId): void
+    {
+    }
```
[the remark this answers is not in the document]
Comment (koriym) [reply src/Log/NullSemanticLogger.php]: @coderabbit Not applicable — src/Log/NullSemanticLogger.php was removed in the 0.9 adaptation; the core Koriym\SemanticLogger\NullSemanticLogger is used instead.

File: src/Log/SafeSemanticLogger.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        return [];
+    }
+
+    /** @param array<string, mixed> $data */
+    public function __unserialize(array $data): void
+    {
+        $this->logger = new SemanticLogger();
+        $this->broken = false;
+    }
```
[the remark this answers is not in the document]
Comment (koriym) [reply src/Log/SafeSemanticLogger.php]: @coderabbit Not applicable — the project's PHPMD config reports no findings and __unserialize()'s signature is fixed by PHP.

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,48 @@ All notable changes to this project will be documented in this file.
...
+- The demos verify themselves: every script prints the semantic log (tree + pretty JSON) and validates the flushed session offline against `docs/schemas/context`, printing a one-line verdict and exiting non-zero on any violation. `demo/run-dependency.php` scenario 3 is command-driven — a PUT on `LevelThree` opens a `command` scope whose purge cascades to level-two/level-one — alongside the manual `manual_purge` entry kind in scenario 6. `demo/run.php` binds real in-memory pools so its log shows genuine cache hits (the `QueryRepositoryModule` default `NullAdapter` made every GET miss).
+- `demo/run-degraded.php`: the failure vocabulary the other demos never reach, in ten sessions — `put_skipped` (all three reasons), a 4xx `command_result` with no invalidation, `save_view`, `manual_store`/`manual_invalidate`, `cache_error` (`read` and `write`), `invalidate` with `cdn` `purged`/`failed` and `roPool`/`etagPool` `failed`, `saved: false` on every save context, `manual_store_result{failed}`, all three `command` producers and a finite `ttl`. With it the demos emit every context type, every schema enum value and every command source the package can produce.
```
[the remark this answers is not in the document]
Comment (koriym) [reply CHANGELOG.md]: @coderabbit Already correct — CHANGELOG.md reads thirteen sessions (the demo gained two scenarios in fd17107).

File: demo/validate.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+{
+    $count = 0;
+    $opens = $node['open'] ?? [];
+    if (is_array($opens)) {
+        foreach ($opens as $child) {
+            if (! is_array($child)) {
+                continue;
+            }
+
+            $count += 1 + countLogEntries($child); // the scope itself + its contents
```
[the remark this answers is not in the document]
Comment (koriym) [reply demo/validate.php]: @coderabbit Fixed in 0e229d2 — widened to array<array-key, mixed>.

File: docs/llms-full.txt [hunk 6]
```diff
@@ -173,52 +173,47 @@ class BlogPage extends ResourceObject
...
+- Turning cache logging off: bind the no-op logger in your application module — `$this->bind(SemanticLoggerInterface::class)->to(NullSemanticLogger::class);` (`Koriym\SemanticLogger\NullSemanticLogger`) — every logging call becomes a no-op.
+- Manual-scope rooting (`manual_store` / `manual_purge` / `manual_invalidate`) is granted by the `TopLevelAwareInterface` capability (`isTopLevel()`), which `SafeSemanticLogger` implements; a custom `SemanticLoggerInterface` decorator can implement it to mark top-level direct calls as application-initiated in its sessions. Without it those calls are recorded as bare root-level events.
+- Flushing is the host's responsibility: the framework only accumulates. Call `flush()` on the injected `SemanticLoggerInterface` (e.g. at request end) to obtain the `LogJson` session for storage or rendering (`vendor/bin/stree`). `flush()` never throws and always resets, so it needs no guard. Under PHP-FPM an unflushed session is simply discarded with the request — the compiled app object unserializes with a fresh session per request (`SafeSemanticLogger::__unserialize()`). Under long-running runtimes, flush per request or bind `NullSemanticLogger`.
```
[the remark this answers is not in the document]
Comment (koriym) [reply docs/llms-full.txt]: @coderabbit Fixed in 0e229d2 — per-request flushing is now stated as valid for sequential worker loops only, with a per-request injector or NullSemanticLogger for concurrent coroutines until #179.

File: src/ResourceStorage.php [hunk 7]
```diff
@@ -327,13 +411,14 @@ private function getVary(): string
...
+        $ttl = $ttl === null ? null : max(0, $ttl);
         $tags = $surrogateKeys !== '' ? explode(' ', $surrogateKeys) : [];
         $tags[] = (new UriTag())($uri);
         /** @var list<string> $uniqueTags */
         $uniqueTags = array_values(array_unique($tags));
-        $this->logger->log('save-etag', ['uri' => (string) $uri, 'etag' => $etag, 'surrogateKeys' => $uniqueTags]);
         // The header value is a quoted entity-tag; the pool key is the bare opaque-tag
-        $this->saver->__invoke(trim($etag, '"'), 'etag', $this->etagPool, $uniqueTags, $ttl);
+        $saved = $this->saver->__invoke(trim($etag, '"'), 'etag', $this->etagPool, $uniqueTags, $ttl);
+        $this->logger->event(new SaveEtagContext((string) $uri, $etag, $uniqueTags, $ttl, $saved));
```
[the remark this answers is not in the document]
Comment (koriym) [reply src/ResourceStorage.php:406]: @coderabbit Fixed in 6a1f5c3, and your comment surfaced the same pattern in DonutRepository's two pre-write cleanup sites, fixed in 0e229d2 — the injected UriTagInterface now reaches all of them.

File: tests/CACHE_DEPENDENCY_TESTS.md [hunk 8]
```diff
@@ -92,6 +223,33 @@ All dependency tests verify both resource cache and ETag invalidation:
...
+  Swoole/RoadRunner a host must flush at the boundary itself. Under *concurrent*
+  coroutines sharing the one singleton, an interleaved open/close violates LIFO —
+  the core logger records the violation as `semantic_logger_error` diagnostics
+  (`close_id_mismatch`, `unclosed_at_flush`) in-band and keeps going, so the
+  interleaved requests' trees are preserved but cross-nested (the misuse is
+  visible, not silent). Cache behavior is unaffected (logging is a side-channel
+  and the core never throws) and every `flush()` resets the session (see
+  `SafeSemanticLoggerTest`). Making the logger request/coroutine-scoped
+  (so concurrent sessions cannot cross-nest or drop) is the robust fix and is
+  intentionally deferred to the host flush-lifecycle work.
```
[the remark this answers is not in the document]
Comment (koriym) [reply tests/CACHE_DEPENDENCY_TESTS.md]: @coderabbit Fixed in 0e229d2 — the limitation now names the consequence: a rejected out-of-order close leaves that request's scope unclosed at flush.

File: tests/RecordingTopLevelAwareLogger.php [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        $this->closes[] = $context;
+        $this->depth--;
+    }
+
+    /** {@inheritDoc} */
+    #[Override]
+    public function flush(array $links = []): LogJson
+    {
+        return new LogJson(self::SEMANTIC_LOG_SCHEMA_URL, [], [], [], $links);
+    }
```
[the remark this answers is not in the document]
Comment (koriym) [reply tests/RecordingTopLevelAwareLogger.php:72]: @coderabbit Fixed in 0e229d2 — flush() resets depth, matching the SafeSemanticLogger contract it mimics.

File: tests/ResourceStorageTest.php [hunk 10]
```diff
@@ -106,4 +181,98 @@ public function testEtagIsNotRegisteredAsInvalidationTag(): void
...
+        // ResourceStorageSaver is final, so the failure is induced one layer down: an inner
+        // pool whose commit() rejects every entry (e.g. storage full).
+        $failingPool = new TagAwareAdapter(new class extends ArrayAdapter {
+            #[Override]
+            public function commit(): bool
+            {
+                return false;
+            }
+        });
+        $poolProvider = new class ($failingPool) implements ProviderInterface{
```
[the remark this answers is not in the document]
Comment (koriym) [reply tests/ResourceStorageTest.php]: @coderabbit Fixed in 0e229d2.

File: docs/llms-full.txt [hunk 11]
```diff
@@ -173,52 +173,49 @@ class BlogPage extends ResourceObject
...
+| `invalidate` (`tags`/`roPool`/`etagPool`/`cdn`/`durationMs`) | event | Tag invalidation outcome; `cdn` is tri-state: `purged` (a configured purger ran), `failed` (the purge threw — fail-closed, after the local pools were invalidated), `skipped` (NullPurger, no CDN configured). Pre-write cleanup vs real invalidation is recorded at the source via the `pre_write_cleanup` marker — see below |
+| `cdn_headers` (`uri`/`headers`/`surrogateKeys`) | event | The CDN-facing headers a donut write or refresh put on the response, read back after they are final: the literal name => value pairs (`CDN-Cache-Control`, `Surrogate-Control`, `Akamai-Cache-Control`, `Surrogate-Key`, `Edge-Cache-Tag`) — including the bound setter's default when no `sMaxAge` was requested. No lifetime header means the CDN serves the page live (the `putDonut` kind). `surrogateKeys` is the split key list a purge's tags must reach to drop the response at the edge. A state-served hit replays the stored headers and emits no new event |
```
[the remark this answers is not in the document]
Comment (koriym) [reply docs/llms-full.txt]: @coderabbit Fixed in 431eb63 — the record now says the response gave the CDN no lifetime directive; the CDN's own behavior stays outside the log. Same correction in llms.txt, what-the-log-proves.md, CACHE_DEPENDENCY_TESTS.md, the context docblock and the CHANGELOG.

File: docs/schemas/context/cdn_headers.json [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    "properties": {
+        "uri": {
+            "type": "string"
+        },
+        "headers": {
+            "type": "object",
+            "additionalProperties": {
+                "type": "string"
+            }
+        },
```
[the remark this answers is not in the document]
Comment (koriym) [reply docs/schemas/context/cdn_headers.json:30]: @coderabbit Fixed in 431eb63 — propertyNames restricts the map to the five managed header names.

File: docs/what-the-log-proves.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+| 3 | Was the CDN told to purge — exactly what, and did it work? | `invalidate` `{tags, roPool, etagPool, cdn, durationMs}`; the purger receives exactly the recorded tags | `cdn` is tri-state (`purged` / `failed` / `skipped`) and fail-closed: local pools first, outcome logged, then the purge exception propagates. One refusing pool is enough to report failure |
+| 4 | Did a conditional request revalidate at the edge? | `conditional_request` `{ifNoneMatch}` closing `cache_hit`/`cache_miss` at layer `etag` — the 304 decision, made before any resource runs | Both `HttpCacheInterface` implementations record it; dropping either recording or swapping the outcome fails the suite |
+| 5 | Why is there no entry — and is a miss cold or degraded? | `put_skipped` `{reason, code}`, `cache_error` `{operation, exceptionClass}` paired with the still-closing `cache_miss` | Skip reasons, the failing side (`read`/`write`) and the throwable class are pinned; `cache_error` + `cache_miss` = outage, lone `cache_miss` = cold |
```
[the remark this answers is not in the document]
Comment (koriym) [reply docs/what-the-log-proves.md]: @coderabbit Fixed in 431eb63 — the row now qualifies the pairing by operation: read.

File: src/CliHttpCache.php [hunk 14]
```diff
@@ -34,7 +44,11 @@ public function isNotModified(array $server): bool
             return false;
         }
 
-        return $this->storage->hasEtag($etag);
+        $openId = $this->logger->open(new ConditionalRequestContext($etag));
+        $hit = $this->storage->hasEtag($etag);
+        $this->logger->close($hit ? new CacheHitContext('etag') : new CacheMissContext('etag'), $openId);
+
+        return $hit;
```
[the remark this answers is not in the document]
Comment (koriym) [reply src/CliHttpCache.php:63]: @coderabbit Fixed in dc85adc (before this pass) — both implementations record cache_error{read} and close cache_miss{etag} on a throwing lookup, then rethrow; pinned for both classes with a throwing ETag pool.

File: src/ResourceBodyEvaluator.php [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        /** @psalm-suppress MixedAssignment $item */
+        foreach ($body as &$item) {
+            if ($item instanceof RequestInterface) {
+                $item = $this->materialize($item);
+            }
+
+            if ($item instanceof ResourceObject) {
+                $item->body = $this($item->body);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (koriym) [reply src/ResourceBodyEvaluator.php]: @coderabbit Fixed in 431eb63 — a ResourceObject sitting in a body now gets its own copy before its body is rewritten, so the shallow clone no longer leaves the live graph mutable one level down. Pinned by a test that fails without the copy.

File: tests/CACHE_DEPENDENCY_TESTS.md [hunk 16]
```diff
@@ -26,6 +28,20 @@ purge(LevelTwo) → LevelOne invalidated
...
+
+A class without `#[Cacheable]` gets no implicit self-purge (no `RefreshSameCommand`),
+so a write only invalidates what its `#[Purge]` attributes name — including its own URI.
+`RefreshInterceptor` is the sole interceptor on this path.
+
+```
+PUT PurgeSrc #[Purge(self)] #[Purge(LevelTwo)]
+  → PurgeSrc's own entry invalidated (nothing else would)
+  → purge(LevelTwo) → LevelOne invalidated
+```
```
[the remark this answers is not in the document]
Comment (koriym) [reply tests/CACHE_DEPENDENCY_TESTS.md]: @coderabbit Fixed in 431eb63 — the fence is marked text.

File: tests/CdnCacheControlHeaderTest.php [hunk 17]
```diff
@@ -28,11 +31,21 @@ public function testCdnCacheControl(): void
         $this->assertArrayHasKey(Header::CDN_CACHE_CONTROL, $ro->headers);
         $this->assertSame($ro->headers[Header::CDN_CACHE_CONTROL], 'max-age=10 stale-while-revalidate=10');
         $repository = $injector->getInstance(QueryRepositoryInterface::class);
+        $logger = $injector->getInstance(SemanticLoggerInterface::class);
         assert($repository->purge(new Uri('page://self/html/comment')));
+        $logger->flush(); // drain the put and the purge: the tree below is the refresh alone
```
[the remark this answers is not in the document]
Comment (koriym) [reply tests/CdnCacheControlHeaderTest.php]: @coderabbit Fixed in 431eb63 — the purge() call is out of assert() and its result asserted, so zend.assertions=-1 can no longer skip the refresh path.

File: tests/DonutCacheTest.php [hunk 18]
```diff
@@ -38,4 +43,136 @@ public function testGetState(): void
...
+    public function testLegacyDonutPayloadCanRecordStorageState(): void
+    {
+        // A warm cache surviving a deploy: a re-saved legacy entry records the lifetime and
+        // tags it is stored with, while the content state it never had stays absent.
+        $donut = $this->legacyDonut()->withStorageState(100, ['fresh-tag']);
+
+        $this->assertSame(['fresh-tag'], $donut->getStorageTags());
+        $this->assertLessThanOrEqual(100, $donut->getRemainingStorageTtl());
+        $this->assertNull($donut->getUnchangedLastModified('tmpl'));
+    }
```
[the remark this answers is not in the document]
Comment (koriym) [reply tests/DonutCacheTest.php:106]: @coderabbit Fixed in 431eb63 — assertIsInt plus a range, so null no longer satisfies the assertion.

File: tests/DonutRepositoryTest.php [hunk 19]
```diff
@@ -134,4 +170,215 @@ public function testInvalidateTags(): void
...
+
+        $setters = [];
+        foreach ((array) glob(dirname(__DIR__) . '/src/{,Cdn/}*.php', GLOB_BRACE) as $file) {
+            $class = 'BEAR\QueryRepository\\' . (str_contains((string) $file, '/Cdn/') ? 'Cdn\\' : '') . basename((string) $file, '.php');
+            if (class_exists($class) && is_subclass_of($class, CdnCacheControlHeaderSetterInterface::class)) {
+                $setters[] = $class;
+            }
+        }
+
+        $this->assertContains(CdnCacheControlHeaderSetter::class, $setters, 'discovery sees the default setter');
```
[the remark this answers is not in the document]
Comment (koriym) [reply tests/DonutRepositoryTest.php]: @coderabbit Fixed in 431eb63 — two explicit globs instead of GLOB_BRACE, and both Cdn/ flavors must be discovered, so the Cdn/ half cannot silently find nothing.

File: tests/QueryRepositoryTest.php [hunk 20]
```diff
@@ -232,4 +257,26 @@ protected function configure(): void
...
         $this->assertInstanceOf(Repository::class, $unserilizedRepository);
     }
+
+    public function testStateWithoutStoredAtDerivesAgeFromLastModified(): void
+    {
+        // A warm cache surviving a deploy holds entries saved before storedAt existed.
+        // Their residence time is unknown, so Age falls back to the content's change time,
+        // which over-states residence rather than presenting a cache hit as fresh.
+        $module = new FakeEtagPoolModule(ModuleFactory::getInstance('FakeVendor\HelloWorld'));
+        $module->override(new class extends AbstractModule{
```
[the remark this answers is not in the document]
Comment (koriym) [reply tests/QueryRepositoryTest.php]: @coderabbit Fixed in 431eb63.

**Comment by koriym** [comment]
@coderabbitai No outstanding blockers — all 20 review comments are answered inline (fixed in 6a1f5c3, 0e229d2, dc85adc, 431eb63, or declined with the reason). Please resolve all open conversations and approve this PR.

The second pass was the strongest one: three tests of mine could not fail (purge() inside assert(), a remaining-TTL assertion null satisfies, a closure guard that proved only half its scan), and the shallow-clone finding was a real invariant hole. All four are now pinned by tests that fail without the fix.

**Comment by koriym** [comment]
@coderabbitai review

Since the last pass this branch changed what happens by default. Recording is now opt-in: `QueryRepositoryModule` binds `NullSemanticLogger` beside its existing `NullAdapter` default, the recording binding moved out of the internal `DonutCacheModule`, and two modules turn it on — `DevQueryRepositoryLogModule` (one file per request, `stree`-readable `latest.json`) and `ProdQueryRepositoryLogModule` (buffer the session, decide at flush: mutations, failures, a sample).

Two decisions are worth your attention:

1. **Flush lives in a sink, not on `transfer()`.** `ShutdownFlush` arms at construction and again at `SafeSemanticLogger::__unserialize()` — the only hook that runs per request, since `configure()` runs once at compile time. `__serialize()` now carries the sink (configuration) while still refusing to carry the session (state). A hook on `ResourceObject::transfer()` would miss both the rendering that happens inside it and the 304 path that never reaches it.
2. **The retention policy is the first consumer of the published reading rules.** `KeepMutationsAndFailures` decides pre-write cleanup vs real invalidation exactly as `docs/llms-full.txt` states it: an `invalidate` is cleanup iff the `pre_write_cleanup` marker immediately precedes it in the same scope.

Please also check the concurrency guard. `HostRuntime` consults mode (`RR_MODE`, or being inside a Swoole coroutine), never capability — the first draft used `extension_loaded('swoole')` and warned on a machine that serves over FPM. Coroutine-local session isolation stays open in #179; it is a separate seam from flush timing.

**Comment by koriym** [comment]
@coderabbitai full review

**Comment by koriym** [comment]
@coderabbitai review

**Comment by koriym** [comment]
For the record: CodeRabbit cannot review this branch any more — `Review skipped: 162 files exceed the limit of 100`, and the incremental command hits the same PR-level diff. The three newest commits (`ebec15c`, `465c63f`, `ebef875`) therefore carry no CodeRabbit pass. The 40 inline comments from its earlier rounds are all resolved.

What that delta did get instead: PHPStan (max) / Psalm (level 1) / phpcs 4.0 / PHPMD clean, 244 tests / 687 assertions green under random order, zero uncovered lines in the new code, mutation checks on the retention rules and the PSR-3 adapter contracts, and the sink driven end-to-end through a real process for both endings it exists for (a request that returns, and a request that `exit()`s like a 304).

**Comment by koriym** [comment]
Two independent reviews ran against the opt-in delta (`0adf2ba..HEAD`) since CodeRabbit cannot: a code-quality pass and a read-only security pass. Both proved their findings by running the code, and between them they found **three defects worth blocking on**. Fixed in `eb1d27f`, Windows follow-up in `1116bd7`, CI 20/20.

**1. The flush was unguarded inside `register_shutdown_function`.** A destination that throws — a wrong `logDir`, a read-only container FS, an unmounted volume — became an uncaught fatal with **exit code 255**, after the response had already been written. That is the opposite of what `DevQueryRepositoryLogModule`'s own docblock promises. The flush is now as total as the core logger it wraps: reported to `error_log()`, swallowed.

**2. The concurrency guard could stop an app from booting.** It raised `trigger_error(E_USER_WARNING)` from `arm()`, which runs while the injector builds the logger. A host with a warning-to-exception handler — more common on RoadRunner and Swoole, not less — got an exception out of the provider. The guard meant to protect those hosts took them down. Diagnostics now use `error_log()`, which no handler can promote.

**3. Refusing to arm only skipped the flush; recording continued.** Nothing else calls `flush()`, so the session grew ~1.4 KB per request for the life of a worker, silent after the first message. `arm()` now returns `false` and `SafeSemanticLogger` falls back to the no-op logger, so "we will not flush here" and "we are not recording here" are one statement — which is what the Prod module's docblock already claimed.

Also fixed, each with a test that fails without it: the null default was prototype-scoped, so `DefaultLogBindingTest` inspected a logger the resource never used and **passed for the wrong reason**; the session filename's nanosecond suffix was not monotonic within a second, so pruning could delete the *newer* session; every filesystem call was unchecked, including two ordinary races (concurrent `mkdir`, concurrent prune); `latest.json` is renamed into place so `stree` cannot open a half-written file; `LogStreamWriter` accepted any wrapper string, so a module argument could truncate an unrelated file via `php://filter` or ship every session to an `ftp://` host, and it dropped the lock on exactly the `php://stdout` default whose docblock promised one; sessions are written 0600 in a 0700 directory, since they carry URIs **with their query strings**, client validators and raw exception text; and the demo pointed at a predictable path in a world-writable directory while telling apps to use their shared `var/log`, which the pruner would have eaten.

Declined, with the reason recorded: a built-in redactor for exception messages and URIs. A regex would earn trust it cannot keep, and the message is the diagnostic. The seam is documented instead — decorate `LogWriterInterface` — and the same decorator is the answer to the retention flood a pool outage causes, since every request then emits `cache_error` and retention rises to 100%.

Now: 260 tests / 719 assertions, zero uncovered lines in the new code, and the mutation checks cover the new invariants (removing the silencing, the flush guard, the stream validation, the file mode, the singleton scope or the ordering fix each fails a test).

File: src/Log/LogFileWriter.php [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+    #[Override]
+    public function write(LogJson $log): void
+    {
+        if ($log->open === [] && $log->events === []) {
+            return; // nothing was recorded: an empty file would claim a request had no cache activity
+        }
+
+        if (! is_dir($this->dir)) {
+            mkdir($this->dir, 0755, true);
+        }
```
[the remark this answers is not in the document]
Comment (koriym) [reply src/Log/LogFileWriter.php]: @coderabbitai Already race-tolerant at `d95c6cb`: `prepareDir()` is `is_dir() || @mkdir() || is_dir()` — the second check is the losing worker — and `prune()` uses `@unlink` with the reason inline. `putLatest()` writes a `bin2hex(random_bytes(4))` temp and renames, so a reader never sees a partial file. Warnings are suppressed rather than emitted because this writer runs at shutdown, where a strict error handler turns one into an uncaught fatal.

File: src/Log/ShutdownFlush.php [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        register_shutdown_function(fn () => $this->flush($logger));
+    }
+
+    #[Override]
+    public function flush(SemanticLoggerInterface $logger): void
+    {
+        $this->writer->write($logger->flush());
+    }
```
[the remark this answers is not in the document]
Comment (koriym) [reply src/Log/ShutdownFlush.php:107]: @coderabbitai Contained since `b119086`: `flush()` wraps the write in `try/catch (Throwable)` and reports through `error_log()`, not `trigger_error()`. The channel matters — arming happens while the injector builds the logger, so a host with a strict error handler would turn a warning there into an exception thrown out of the provider, taking down the runtime the sink declines to serve. The class docblock states this.

File: src/ProdQueryRepositoryLogModule.php [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+    ) {
+        parent::__construct($module);
+    }
+
+    #[Override]
+    protected function configure(): void
+    {
+        $this->bind(LogWriterInterface::class)->toInstance(
+            new PolicyLogWriter(new KeepMutationsAndFailures($this->sampleRate), new LogStreamWriter($this->stream)),
+        );
```
[the remark this answers is not in the document]
Comment (koriym) [reply src/ProdQueryRepositoryLogModule.php]: @coderabbitai Valid, fixed in `d1af066`. The docblock promised a seam that did not exist: `configure()` built `KeepMutationsAndFailures` inline and bound the writer with `toInstance`, so nothing could override it. Now `RetentionPolicyInterface` is bound and `PolicyLogWriter` is constructed by the injector, so a binding installed after the module reaches the writer. The inner destination is qualified with a new `#[LogDestination]` attribute because the decorator implements the same interface, following the `#[EtagPool]` convention. `LogWriterTest::testAnAppPolicyBoundAfterTheModuleDecidesInstead` binds an always-keep policy and asserts a healthy read — the session the default drops — reaches the destination; it fails against the previous wiring. CHANGELOG wording corrected too: the override is process-wide, not per request.

**Comment by koriym** [comment]
@coderabbitai review

Three threads answered at `d1af066`. One was valid: `ProdQueryRepositoryLogModule` documented an override seam it did not have, so `RetentionPolicyInterface` is now bound and `PolicyLogWriter` is injector-constructed, with a test that fails against the previous wiring. The other two — the `mkdir`/`prune` races and the shutdown-flush containment — were already in place; the replies point at the code and say why `error_log()` is the channel rather than `trigger_error()`.

**Comment by koriym** [comment]
@coderabbitai Thanks — all three threads are resolved and your check of `d1af066` matches mine locally (phpcs, PHPStan max, Psalm level 1, PHPMD, 266 tests, four demos exit 0). The PR still carries your `CHANGES_REQUESTED` from 2026-08-15, which predates the fixes. Could you submit an approving review so the state reflects the resolved threads?
